aptos-move/e2e-move-tests/src/tests/public_structs_enums_upgrade.rs

```diff
@@ -0,0 +1,213 @@
+// Copyright (c) Aptos Foundation
+// Licensed pursuant to the Innovation-Enabling Source Code License, available at https://github.com/aptos-labs/aptos-core/blob/main/LICENSE
+
+//! Tests for upgrade compatibility of public/package structs/enums
+
+use crate::{assert_success, assert_vm_status, MoveHarness};
+use aptos_framework::BuildOptions;
+use aptos_language_e2e_tests::account::Account;
+use aptos_package_builder::PackageBuilder;
+use aptos_types::{account_address::AccountAddress, transaction::TransactionStatus};
+use move_core_types::vm_status::StatusCode;
+
+#[test]
+fn public_enum_upgrade() {
+    let mut h = MoveHarness::new();
+    let acc = h.new_account_at(AccountAddress::from_hex_literal("0x815").unwrap());
+
+    let result = publish(
+        &mut h,
+        &acc,
+        r#"
+        module 0x815::m {
+            enum Data {
+               V1{x: u64}
+            }
+        }
+    "#,
+    );
+    assert_success!(result);
+
+    let result = publish(
```

Comment:
Looks like this is testing both upgrade to `public` and adding a new variant. Can we do that in two steps (first upgrade to `public` without any changes to the enum, next upgrade the `public` enum to have a new variant)?